compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -548,21 +556,244 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
+    /// LL |         let y: &dyn Foo = x;
+    ///    |                           ^ cast requires that `'a` must outlive `'static`
+    /// LL |         y.hello();
+    ///    |         --------- calling this method introduces a `'static` lifetime requirement
+    ///    |
+    /// note: the `impl` on `(dyn a::Foo + 'static)` has a `'static` lifetime requirement
+    ///   --> $DIR/static-impl-obligation.rs:4:10
+    ///    |
+    /// LL |     impl dyn Foo {
+    ///    |          ^^^^^^^
+    /// help: relax the implicit `'static` bound on the impl
+    ///    |
+    /// LL |     impl dyn Foo + '_ {
+    ///    |                  ++++
+    /// ```
+    /// ```text
+    /// error: lifetime may not live long enough
+    ///   --> $DIR/static-impl-obligation.rs:173:27
+    ///    |
+    /// LL |     fn bar<'a>(x: &'a &'a u32) {
+    ///    |            -- lifetime `'a` defined here
+    /// LL |         let y: &dyn Foo = x;
+    ///    |                           ^ cast requires that `'a` must outlive `'static`
+    /// LL |         y.hello();
+    ///    |         --------- calling this method introduces a `'static` lifetime requirement
+    ///    |
+    /// note: the `impl` on `(dyn p::Foo + 'static)` has `'static` lifetime requirements
+    ///   --> $DIR/static-impl-obligation.rs:169:20
+    ///    |
+    /// LL |     impl dyn Foo + 'static where Self: 'static {
+    ///    |                    ^^^^^^^             ^^^^^^^
+    /// LL |         fn hello(&self) where Self: 'static {}
+    ///    |                                     ^^^^^^^
+    /// ```
+    #[allow(rustc::diagnostic_outside_of_impl)]
+    #[allow(rustc::untranslatable_diagnostic)]
+    fn explain_impl_static_obligation(
+        &self,
+        diag: &mut Diag<'_>,
+        ty: Ty<'tcx>,
+        call_span: Span,
+        outlived_fr: RegionVid,
+    ) {
+        let tcx = self.infcx.tcx;
+        let ty::FnDef(def_id, args) = ty.kind() else {
+            return;
+        };
+        let Ok(Some(instance)) = ty::Instance::resolve(
+            tcx,
+            self.param_env,
+            *def_id,
+            self.infcx.resolve_vars_if_possible(args),
+        ) else {
+            return;
+        };
+        debug!(?instance);
+        let def_id = instance.def_id();
+        let mut parent = tcx.parent(def_id);
+        debug!(?def_id, ?parent);
+        let trait_preds: Vec<_> = match tcx.def_kind(parent) {
+            hir::def::DefKind::Impl { .. } => tcx
+                .trait_id_of_impl(parent)
+                .map_or(vec![], |id| elaborate_predicates_of(tcx, id).collect()),
+            hir::def::DefKind::Trait => {
+                let Some(ty) = args.get(0).and_then(|arg| arg.as_type()) else {
+                    return;
+                };
+                let mut impls = vec![];
+                tcx.for_each_relevant_impl(parent, ty, |id| {
+                    impls.push(id);
+                });
+                if let [def_id] = impls[..] {
```

Comment:
This won't work if there's a built-in impl that applies, and I kind of prefer that we remove this hack. If `Instance::resolve` doesn't give you the right item back, it's probably because there's actual polymorphism going on here.